server/src-lib/Hasura/RQL/Types/Permission.hs

```diff
@@ -66,6 +68,13 @@ newtype UserVars
   = UserVars { unUserVars :: Map.HashMap SessVar SessVarVal}
   deriving (Show, Eq, FromJSON, ToJSON, Hashable)
 
+-- | We need this to transform the underlying session variable map
+-- as the UserVars constructor is not exposed
+mapUserVars
+  :: (Map.HashMap SessVar SessVarVal -> Map.HashMap SessVar SessVarVal)
+  -> UserVars -> UserVars
+mapUserVars f = UserVars . f . unUserVars
```

Comment:
This seems like it somewhat defeats the purpose of having the smart constructor, since one can trivially defeat the intended invariants by writing `mapUserVars (const myEvilHashMap) (mkUserVars [])`. Of course, I do not expect anyone would actually do that, but they might end up violating the invariants by mistake in a more complicated way.

Perhaps define a more restrictive `filterUserVars`, instead? Or use `mkUserVars` to reconstruct the result instead of the `UserVars` constructor?